lib/cli/src/Cardano/CLI.hs

```diff
@@ -227,164 +215,118 @@ import qualified Data.Text.Encoding as T
+        , "an active server and can be run offline (e.g. 'mnemonic generate')."
+        ])
 
 {-------------------------------------------------------------------------------
-                              CLI Construction
+                            Commands - 'mnemonic'
+
+  cardano-wallet mnemonic generate [--size=INT]
 -------------------------------------------------------------------------------}
 
--- | Make a specialized CLI from the specified commands, options, and examples.
---
-makeCli :: String -> String -> String -> Docopt
-makeCli cliCommandsSpecific cliOptionsSpecific cliExamplesSpecific =
-    fromRight (error "Unable to construct CLI.") $
-        parseUsage usageString
+-- | cardano-wallet mnemonic
+cmdMnemonic :: Mod CommandFields (IO ())
+cmdMnemonic = command "mnemonic" $ info (helper <*> cmds) mempty
   where
-    usageString = mempty
-        <> cliHeader
-        <> "\nUsage:\n"
-        <> cliCommands
-        <> "\nOptions:\n"
-        <> cliOptions
-        <> "\nExamples:\n"
-        <> cliExamples
-    cliCommands = unlines $ filter (not . null) $ lines $ mempty
-        <> cliCommandsSpecific
-        <> cliCommandsGeneric
-    cliOptions = unlines $ filter (not . null) $ sort $ lines $ mempty
-        <> cliOptionsSpecific
-        <> cliOptionsGeneric
-    cliExamples = unlines $ filter (not . null) $ lines $ mempty
-        <> cliExamplesSpecific
-        <> cliExamplesGeneric
-
-cliHeader :: String
-cliHeader = [here|Cardano Wallet CLI.
-
-The CLI is a proxy to the wallet server, which is required for most
-commands. Commands are turned into corresponding API calls, and submitted
-to an up-and-running server. Some commands do not require an active server
-and can be run "offline". (e.g. 'generate mnemonic')
-
-    ⚠️  Options are positional (--a --b is not equivalent to --b --a) ! ⚠️
-|]
-
-cliCommandsGeneric :: String
-cliCommandsGeneric = [here|
-  cardano-wallet mnemonic generate [--size=INT]
+    cmds = subparser $ mempty
+        <> cmdMnemonicGenerate
+
+-- | Arguments for 'mnemonic generate' command
+newtype MnemonicGenerateArgs = MnemonicGenerateArgs
+    { _size :: Int
+    }
+
+-- | cardano-wallet mnemonic generate [--size=INT]
+cmdMnemonicGenerate :: Mod CommandFields (IO ())
+cmdMnemonicGenerate = command "generate" $ info (helper <*> cmd) $ mempty
+    <> progDesc "Generate English BIP-0039 compatible mnemonic words."
+  where
+    cmd = exec . MnemonicGenerateArgs <$> sizeOption
+    exec (MnemonicGenerateArgs n) = do
+        m <- case n of
+            9  -> mnemonicToText @9 . entropyToMnemonic <$> genEntropy
+            12 -> mnemonicToText @12 . entropyToMnemonic <$> genEntropy
+            15 -> mnemonicToText @15 . entropyToMnemonic <$> genEntropy
+            18 -> mnemonicToText @18 . entropyToMnemonic <$> genEntropy
+            21 -> mnemonicToText @21 . entropyToMnemonic <$> genEntropy
+            24 -> mnemonicToText @24 . entropyToMnemonic <$> genEntropy
+            _  -> do
```

Comment:
I know this code already existed, but I wonder if there's a way to remove the duplication here?

Reply:
Not _really_. We could reduce the effort with a type-class but it'd make readability worse IMO. The trick here is that we are working with different types on each branch, and we flatten them out to `Text` with `mnemonicToText`. Because they're all different types, it's hard to handle them in a uniform way unless we rely on ad-hoc polymorphism. 

Not sure it's worth the effort to be honest, we are better off a few similar but very understandable lines :)

Side-note: as an example of how this could be handled in a more polymorphic way, have a look at: https://github.com/input-output-hk/cardano-wallet/blob/master/lib/core/src/Cardano/Wallet/Primitive/AddressDerivation.hs#L279-L358

Reply:
> Not sure it's worth the effort to be honest, we are better off a few similar but very understandable lines :)

Fair enough! I had a go at this. Doing so does indeed end up introducing a lot of class constraints, which ends up longer than the original. :man_shrugging: